### PR TITLE
Fix for change in material and render-data properties with batching

### PR DESCRIPTION
**Pull request by roshanch**
1. Creates a pool of batches which is created initially, batches are taken from this pool.
2. Adding event listener for renderdata as part of mesh, material, render-pass which can be shared by multiple render-data
3. Disables batching for the render-data if mesh, material or any of the property changes in OnStep().

Tested with jassimp, gvr-testmodels, gvr-tutorial-6, litshader, outline sample, gvr-video, gvr-controller, planetarium and custom sample where light, material, mesh is modified in onStep().

GearVRf-DCO-1.0-Signed-off-by: Roshan Chaudhari roshan.c@samsung.com

### CODE REVIEW

File: GVRf/Framework/framework/src/main/jni/engine/renderer/batch.cpp [hunk 1]
```diff
@@ -164,68 +160,52 @@ bool Batch::isRenderDataDisabled(){
...
-    bool update_vbo = isRenderDataDisabled();
-
-    if(render_data_set_.size() ==0)
-        return;
-
-    if(update_vbo)
-        regenerateMeshData();
+void Batch::resetBatch(){
+    clearData();
+    delete renderdata_;
```
Comment (liaxim) [inline GVRf/Framework/framework/src/main/jni/engine/renderer/batch.cpp]: Wouldn't it be safer if renderdata_ was also set to nullptr? In the future the dtor may actually execute and the flow may be such that this renderdata_ may be deleted second time.

File: GVRf/Framework/framework/src/main/jni/engine/renderer/batch.cpp [hunk 2]
```diff
@@ -164,68 +160,52 @@ bool Batch::isRenderDataDisabled(){
...
+    Renderer::batch_manager->deallocateBatch(this);
```
Comment (liaxim) [inline GVRf/Framework/framework/src/main/jni/engine/renderer/batch.cpp]: You may consider a more canonical name for a pool like "free". deallocate could create a wrong impression.

File: GVRf/Framework/framework/src/main/java/org/gearvrf/GVRMesh.java [hunk 3]
```diff
@@ -362,14 +362,6 @@ public GVRMesh getBoundingBox() {
         return mBones;
     }
     /**
```
Comment (liaxim) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/GVRMesh.java]: So what change makes the need to distinguish static vs dynamic mesh not needed anymore? How is the dynamic case handled now?

File: GVRf/Framework/framework/src/main/jni/engine/renderer/batch.h [hunk 4]
```diff
@@ -20,13 +20,13 @@
 #include "objects/components/render_data.h"
 #include <map>
```
Comment (liaxim) [inline GVRf/Framework/framework/src/main/jni/engine/renderer/batch.h]: This file uses tabs mostly for indentation; please fix it at some point.

**Comment by liaxim** [comment]
Looks good and I can't find a sample broken by this change. Will merge in the afternoon if there are no other comments.

--
Having said all that I do think we need to start the unit tests for the batching. The sooner the better as it will be easier to write and expand them as batching grows in complexity. The fact the samples work doesn't mean that much - there is a lot of branching that the samples will not cover. Also we can't really tell how effective batching is by just looking at the samples. Not to mention ensuring memory management is correct. It is impossible code this complex not to have bugs and it will be much harder for us to find and deal with those bugs as they are discovered in apps by us or end users.

File: GVRf/Framework/framework/src/main/java/org/gearvrf/GVRMesh.java (hunk 3 above)
Comment (roshanch) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/GVRMesh.java]: now any if any of the mesh changes, batching will be disabled for that mesh. so there is no more "dynamic"/ "static" now. same applies for material/ renderdata.

File: GVRf/Framework/framework/src/main/jni/engine/renderer/batch.cpp (hunk 2 above)
Comment (roshanch) [inline GVRf/Framework/framework/src/main/jni/engine/renderer/batch.cpp]: Will change that.

File: GVRf/Framework/framework/src/main/jni/engine/renderer/batch.h (hunk 4 above)
Comment (roshanch) [inline GVRf/Framework/framework/src/main/jni/engine/renderer/batch.h]: ok, I need to change my formatting in studio. will fix it.

**Comment by roshanch** [comment]
Sure.! So I have sample which I modified for different cases for the batching. I will include that in Nola's test suit.

File: GVRf/Framework/framework/src/main/jni/engine/renderer/batch.cpp (hunk 1 above)
Comment (roshanch) [inline GVRf/Framework/framework/src/main/jni/engine/renderer/batch.cpp]: okay. will make that change.

**Pull request merged**